doc/developer/platform/formalism.md

```diff
@@ -13,57 +13,466 @@ It does not intend to describe the **current behavior** of the system.
+arbitrary typed data. For example, one might use a piece of data to represent a
+single row in an SQL database, or a single key-value pair in a key-value store.
+
+In this document, we often denote a piece of data as `data`. However, we want
+to emphasize that as far as updates are concerned, each piece of data is an
+opaque, atomic object rather than a composite type. When we describe an update
+to a row, that data includes the entire new row, not just the columns inside it
+which changed.
+
+We say "piece of data" instead of "datum" because "datum" has a different
+meaning in Materialize's internals: it refers to a cell within a row. In the
+CIDR differential dataflow paper, a piece of data is called a "record".
+
+## Times
+
+A *Timeline* is a unique ID (for disambiguating e.g. two sets of integers), a
+set of *Materialize times* `T`, together with a *meet* (greatest lower bound)
+and *join* (least upper bound), each of which take two elements of `T` and
+yield a third:
+
+```
+meet(t1, t2) => t3
+join(t1, t2) => t3
+```
+
+Both `meet` and `join` must be associative, commutative, and idempotent. Each
+timeline therefore forms a
+[lattice](https://en.wikipedia.org/wiki/Lattice_(order)).
+
+A *time* is an element of `T`. We define a partial order `<=t` over times in
+the usual way for join semilattices: for any two times `t1` and `t2` in `T`,
+`t1 <=t t2` iff `join(t1, t2) = t2`. We also define a strict partial order over
+times `<t`, such that `t1 < t2` iff `t1 <= t2` and `t1 != t2`.
+
+Each `T` contains a minimal time `t0` such that `t0 <=t t` for all `t` in `T`.
+Each timeline is therefore [well-founded](https://en.wikipedia.org/wiki/Well-founded_relation).
+
+As a concrete example, we might choose `T` to be the naturals, with `max` as
+our join and `t0 = 0`. The usual senses of `<=` and `<` then serve as
+`<=t` and `<t`, respectively.
+
+A *frontier* is an antichain of times in `T`: a set where no two times are
+comparable via `<=t`. We use frontiers to specify a lower (or upper) bound on
+times in `T`. We say that a time `t` *is later than* or *after* a frontier `F`
+iff there exists some `f` in `F` such that `f <t t`. We write this `F <t t`.
+Similarly we can speak of a time `t` which is later than or equal to `F`, by
+which we mean there exists some `f` in `F` such that `f <=t t`: we write this
+`F <=t t`.
+
+For integers, a frontier could be the singleton set `F = {5}`; the set of times
+after `F` are `{6, 7, 8, ...}`.
+
+The latest frontier is the empty set `{}`. No time is later than `{}`.
+
+Throughout this document, "time" generally refers to a Materialize time in some
+timeline. We say "wall-clock time" to describe time in the real world. Times
+from external systems may be
+[reclocked](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20210714_reclocking.md)
+into Materialize times.
+
+## Collection Versions
+
+A *collection version* is a [multiset](https://en.wikipedia.org/wiki/Multiset)
```

Comment:
> The definition from that Wikipedia article doesn't match the one you gave here, since it requires that the multiplicity of elements be nonnegative. I don't think there is a generalized term for what you're describing here; a bit of searching suggests that generalized set and hybrid multiset have been used.

Multisets with negative multiplicities have a fairly established track record! Hailperin 1976 calls them "signed multisets", as opposed to "unsigned multisets". Loeb, as you note, calls them "hybrid multisets". "Generalized Set" is from Whitney (1933). Rado (1975) used signed cardinals for multisets, and Reisig (1985) defines multisets as functions to Z. I'll cite Reisig here; it seems simplest.

> Another thing that comes to mind here is that we haven't defined "type" anywhere, but we refer to that concept here.

If you'd like to suggest language I'd love to hear it! The formalism originally talked about types but didn't really define them, and I wasn't quite sure what to say. I'm not sure "all of the collection versions of a given collection have the same domain" helps a whole lot, because we can always define that domain as the union of the individual collection version domains, and it also doesn't constrain the records *within* a single collection version. I kind of got the sense that the original formalism was grasping towards something (e.g.) like an SQL schema such that each piece of data represented a row with well-defined columns, but perhaps this is a question for @frankmcsherry and a future PR?